M5 Forecasting - Accuracy

Note: This is one of the two complementary competitions that together comprise the M5 forecasting challenge. Can you estimate, as precisely as possible, the point forecasts of the unit sales of various products sold in the USA by Walmart? If you are interested in estimating the uncertainty distribution of the realized values of the same series, be sure to check out its companion competition

How much camping gear will one store sell each month in a year? To the uninitiated, calculating sales at this level may seem as difficult as predicting the weather. Both types of forecasting rely on science and historical data. While a wrong weather forecast may result in you carrying around an umbrella on a sunny day, inaccurate business forecasts could result in actual or opportunity losses. In this competition, in addition to traditional forecasting methods you’re also challenged to use machine learning to improve forecast accuracy.

The Makridakis Open Forecasting Center (MOFC) at the University of Nicosia conducts cutting-edge forecasting research and provides business forecast training. It helps companies achieve accurate predictions, estimate the levels of uncertainty, avoiding costly mistakes, and apply best forecasting practices. The MOFC is well known for its Makridakis Competitions, the first of which ran in the 1980s.

In this competition, the fifth iteration, you will use hierarchical sales data from Walmart, the world’s largest company by revenue, to forecast daily sales for the next 28 days. The data, covers stores in three US States (California, Texas, and Wisconsin) and includes item level, department, product categories, and store details. In addition, it has explanatory variables such as price, promotions, day of the week, and special events. Together, this robust dataset can be used to improve forecasting accuracy.

If successful, your work will continue to advance the theory and practice of forecasting. The methods used can be applied in various business areas, such as setting up appropriate inventory or service levels. Through its business support and training, the MOFC will help distribute the tools and knowledge so others can achieve more accurate and better calibrated forecasts, reduce waste and be able to appreciate uncertainty and its risk implications.

Evaluation:
This competition uses a Weighted Root Mean Squared Scaled Error (RMSSE). Extensive details about the metric, scaling, and weighting can be found in the [M5 Participants Guide](https://mofc.unic.ac.cy/m5-competition/).

In [1]:
import gc
import pandas as pd
import numpy as np

In [2]:
'''
Function to reduce memory usage.
From: https://www.kaggle.com/ragnar123/very-fst-model
'''
def reduce_mem_usage(df, verbose=True):
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    start_mem = df.memory_usage().sum() / 1024**2    
    for col in df.columns:
        col_type = df[col].dtypes
        if col_type in numerics:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)    
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose: print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
    return df

In [3]:
#Load data.
calendar = pd.read_csv('calendar.csv')
calendar = reduce_mem_usage(calendar)

#verify how many weeks do we have. This will help us to analise the prices.
print('weeks: ', len(calendar['wm_yr_wk'].drop_duplicates()))
print()

#Verify null
print('Verify null:')
print(calendar.isnull().sum())
print()
print(calendar.isnull().sum() / len(calendar))
print()
calendar.head()

Mem. usage decreased to  0.12 Mb (41.9% reduction)
weeks:  282

Verify null:
date               0
wm_yr_wk           0
weekday            0
wday               0
month              0
year               0
d                  0
event_name_1    1807
event_type_1    1807
event_name_2    1964
event_type_2    1964
snap_CA            0
snap_TX            0
snap_WI            0
dtype: int64

date            0.000000
wm_yr_wk        0.000000
weekday         0.000000
wday            0.000000
month           0.000000
year            0.000000
d               0.000000
event_name_1    0.917725
event_type_1    0.917725
event_name_2    0.997461
event_type_2    0.997461
snap_CA         0.000000
snap_TX         0.000000
snap_WI         0.000000
dtype: float64



date  wm_yr_wk    weekday  wday  month  year    d event_name_1  \
0  2011-01-29     11101   Saturday     1      1  2011  d_1          NaN   
1  2011-01-30     11101     Sunday     2      1  2011  d_2          NaN   
2  2011-01-31     11101     Monday     3      1  2011  d_3          NaN   
3  2011-02-01     11101    Tuesday     4      2  2011  d_4          NaN   
4  2011-02-02     11101  Wednesday     5      2  2011  d_5          NaN   

  event_type_1 event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0          NaN          NaN          NaN        0        0        0  
1          NaN          NaN          NaN        0        0        0  
2          NaN          NaN          NaN        0        0        0  
3          NaN          NaN          NaN        1        1        0  
4          NaN          NaN          NaN        1        0        1

File 1: “calendar.csv”

Contains information about the dates the products are sold.

     date: The date in a “y-m-d” format.

     wm_yr_wk: The id of the week the date belongs to.
    
     weekday: The type of the day (Saturday, Sunday, …, Friday).
    
     wday: The id of the weekday, starting from Saturday.
    
     month: The month of the date.
    
     year: The year of the date.
    
     event_name_1: If the date includes an event, the name of this event.
    
     event_type_1: If the date includes an event, the type of this event.
    
     event_name_2: If the date includes a second event, the name of this event.
    
     event_type_2: If the date includes a second event, the type of this event.
    
     snap_CA, snap_TX, and snap_WI: A binary variable (0 or 1) indicating whether the stores of CA, TX or WI allow SNAP2 purchases on the examined date. 1 indicates that SNAP purchases are allowed.

In [4]:
sales_train_validation = pd.read_csv('sales_train_validation.csv')
sales_train_validation = reduce_mem_usage(sales_train_validation)

#Verify the quantity of items and store. This will help us to see how many weeks we do not have any prices for a product.
print('Items and stores:', len(sales_train_validation[['item_id', 'store_id']].drop_duplicates()))
print()

#Verify null
print('Verify null:')
print(sales_train_validation.isnull().sum()[sales_train_validation.isnull().sum() > 0]) #there is not any null values.
print()
print(sales_train_validation.isnull().sum()[sales_train_validation.isnull().sum() > 0] / len(sales_train_validation))
print()
sales_train_validation.head()

Mem. usage decreased to 95.00 Mb (78.7% reduction)
Items and stores: 30490

Verify null:
Series([], dtype: int64)

Series([], dtype: float64)



id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  d_1  d_2  d_3  d_4  ...  d_1904  d_1905  d_1906  d_1907  d_1908  \
0       CA    0    0    0    0  ...       1       3       0       1       1   
1       CA    0    0    0    0  ...       0       0       0       0       0   
2       CA    0    0    0    0  ...       2       1       2       1       1   
3       CA    0    0    0    0  ...       1       0       5       4       1   
4       CA    0    0    0    0  ...       2       1       1       0       1   

   d_1909  d_1910  d_1911  d_1912  d_1913  
0       1       3       0       1       1  
1       1       0       0       0       0  
2       1       0       1       1       1  
3       0       1       3       7       2  
4       1       2       2       2       4  

[5 rows x 1919 columns]

File 3: “sales_train.csv”

Contains the historical daily unit sales data per product and store.

     item_id: The id of the product.

     dept_id: The id of the department the product belongs to.

     cat_id: The id of the category the product belongs to.

     store_id: The id of the store where the product is sold.

     state_id: The State where the store is located.

     d_1, d_2, …, d_i, … d_1941: The number of units sold at day i, starting from 2011-01-29.
    
    
We have 282 and 30,490 differents items. Therefore, the ideal would be to have 8,598,180 (282 x 30,400) prices. 

In [5]:
sell_prices = pd.read_csv('sell_prices.csv')
sell_prices = reduce_mem_usage(sell_prices)

print("len: ", len(sell_prices))
print()
#Verify null
print('Verify null:')
print(sell_prices.isnull().sum())
print()
print(sell_prices.isnull().sum() / len(sell_prices))
print()
display(sell_prices.head())
print()

Mem. usage decreased to 130.48 Mb (37.5% reduction)
len:  6841121

Verify null:
store_id      0
item_id       0
wm_yr_wk      0
sell_price    0
dtype: int64

store_id      0.0
item_id       0.0
wm_yr_wk      0.0
sell_price    0.0
dtype: float64



store_id        item_id  wm_yr_wk  sell_price
0     CA_1  HOBBIES_1_001     11325    9.578125
1     CA_1  HOBBIES_1_001     11326    9.578125
2     CA_1  HOBBIES_1_001     11327    8.257812
3     CA_1  HOBBIES_1_001     11328    8.257812
4     CA_1  HOBBIES_1_001     11329    8.257812

File 2: “sell_prices.csv”

Contains information about the price of the products sold per store and date.

     store_id: The id of the store where the product is sold.

     item_id: The id of the product.

     wm_yr_wk: The id of the week.

     sell_price: The price of the product for the given week/store. The price is provided per week (average across seven days). If not available, this means that the product was not sold during the examined week. Note that although prices are constant at weekly basis, they may change through time (both training and test set).
    
    
Considering that we have only 6,841,121 prices, there are  1,757,059 (8,598,180 - 6,841,121) products that were not solds in a given week.

In [6]:
#Define 2 holdout
holdout1 = pd.read_csv('sample_submission.csv')
holdout1 = reduce_mem_usage(holdout1)
holdout2 = pd.read_csv('sample_submission.csv')
holdout2 = reduce_mem_usage(holdout2)


#One holdout is related to forecast between day 1914 and 1941.
#The second, between day 1942 and 1969
test1 = holdout1.copy()
test2 = holdout2.copy()
test1.columns = ['id', 'd_1914', 'd_1915', 'd_1916', 'd_1917', 'd_1918', 'd_1919', 'd_1920', 'd_1921', 'd_1922', 'd_1923',
                    'd_1924', 'd_1925', 'd_1926', 'd_1927', 'd_1928', 'd_1929', 'd_1930', 'd_1931', 
                    'd_1932', 'd_1933', 'd_1934', 'd_1935', 'd_1936', 'd_1937', 'd_1938', 'd_1939', 'd_1940', 'd_1941']
test2.columns = ['id', 'd_1942', 'd_1943', 'd_1944', 'd_1945', 'd_1946', 'd_1947', 'd_1948', 'd_1949', 'd_1950', 'd_1951', 
                    'd_1952', 'd_1953', 'd_1954', 'd_1955', 'd_1956', 'd_1957', 'd_1958', 'd_1959', 
                    'd_1960', 'd_1961', 'd_1962', 'd_1963', 'd_1964', 'd_1965', 'd_1966', 'd_1967', 'd_1968', 'd_1969']

print('Shape test1: ' + str(test1.shape))
print('Shape test2: ' + str(test2.shape))
display(test1.head())
display(test2.head())

Mem. usage decreased to  2.09 Mb (84.5% reduction)
Mem. usage decreased to  2.09 Mb (84.5% reduction)
Shape test1: (60980, 29)
Shape test2: (60980, 29)


id  d_1914  d_1915  d_1916  d_1917  d_1918  \
0  HOBBIES_1_001_CA_1_validation       0       0       0       0       0   
1  HOBBIES_1_002_CA_1_validation       0       0       0       0       0   
2  HOBBIES_1_003_CA_1_validation       0       0       0       0       0   
3  HOBBIES_1_004_CA_1_validation       0       0       0       0       0   
4  HOBBIES_1_005_CA_1_validation       0       0       0       0       0   

   d_1919  d_1920  d_1921  d_1922  ...  d_1932  d_1933  d_1934  d_1935  \
0       0       0       0       0  ...       0       0       0       0   
1       0       0       0       0  ...       0       0       0       0   
2       0       0       0       0  ...       0       0       0       0   
3       0       0       0       0  ...       0       0       0       0   
4       0       0       0       0  ...       0       0       0       0   

   d_1936  d_1937  d_1938  d_1939  d_1940  d_1941  
0       0       0       0       0       0       0  
1       0       0       0       0       0       0  
2       0       0       0       0       0       0  
3       0       0       0       0       0       0  
4       0       0       0       0       0       0  

[5 rows x 29 columns]

id  d_1942  d_1943  d_1944  d_1945  d_1946  \
0  HOBBIES_1_001_CA_1_validation       0       0       0       0       0   
1  HOBBIES_1_002_CA_1_validation       0       0       0       0       0   
2  HOBBIES_1_003_CA_1_validation       0       0       0       0       0   
3  HOBBIES_1_004_CA_1_validation       0       0       0       0       0   
4  HOBBIES_1_005_CA_1_validation       0       0       0       0       0   

   d_1947  d_1948  d_1949  d_1950  ...  d_1960  d_1961  d_1962  d_1963  \
0       0       0       0       0  ...       0       0       0       0   
1       0       0       0       0  ...       0       0       0       0   
2       0       0       0       0  ...       0       0       0       0   
3       0       0       0       0  ...       0       0       0       0   
4       0       0       0       0  ...       0       0       0       0   

   d_1964  d_1965  d_1966  d_1967  d_1968  d_1969  
0       0       0       0       0       0       0  
1       0       0       0       0       0       0  
2       0       0       0       0       0       0  
3       0       0       0       0       0       0  
4       0       0       0       0       0       0  

[5 rows x 29 columns]

In [7]:
#Transpose sales_train_validation so the days goes to rows.
sales_train_validation = pd.melt(sales_train_validation, id_vars = ['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id'], var_name = 'day', value_name = 'demand')

#Verify null
print(sales_train_validation.isnull().sum())
print()
print(sales_train_validation.isnull().sum() / len(sales_train_validation))
print()
display(sales_train_validation.head())
print('...')
display(sales_train_validation.tail())

id          0
item_id     0
dept_id     0
cat_id      0
store_id    0
state_id    0
day         0
demand      0
dtype: int64

id          0.0
item_id     0.0
dept_id     0.0
cat_id      0.0
store_id    0.0
state_id    0.0
day         0.0
demand      0.0
dtype: float64



id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  day  demand  
0       CA  d_1       0  
1       CA  d_1       0  
2       CA  d_1       0  
3       CA  d_1       0  
4       CA  d_1       0

...


id      item_id  dept_id cat_id store_id  \
58327365  FOODS_3_823_WI_3_validation  FOODS_3_823  FOODS_3  FOODS     WI_3   
58327366  FOODS_3_824_WI_3_validation  FOODS_3_824  FOODS_3  FOODS     WI_3   
58327367  FOODS_3_825_WI_3_validation  FOODS_3_825  FOODS_3  FOODS     WI_3   
58327368  FOODS_3_826_WI_3_validation  FOODS_3_826  FOODS_3  FOODS     WI_3   
58327369  FOODS_3_827_WI_3_validation  FOODS_3_827  FOODS_3  FOODS     WI_3   

         state_id     day  demand  
58327365       WI  d_1913       1  
58327366       WI  d_1913       0  
58327367       WI  d_1913       0  
58327368       WI  d_1913       3  
58327369       WI  d_1913       0

In [8]:
#Concat our train database with the holdouts.
#Considering that we do not have any information for days after 1913, this wil not imply in any kind of information leak. 

#Get information about the product id.
product_id = sales_train_validation[['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id']].drop_duplicates()

# merge with product table
test1['id'] = test1['id'].str.replace('_evaluation','_validation')
test1 = test1.merge(product_id, how = 'left', on = 'id')
test2['id'] = test2['id'].str.replace('_evaluation','_validation')
test2 = test2.merge(product_id, how = 'left', on = 'id')

#Transpose so the days goes to rows.
test1 = pd.melt(test1, id_vars = ['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id'], var_name = 'day', value_name = 'demand')
test2 = pd.melt(test2, id_vars = ['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id'], var_name = 'day', value_name = 'demand')

#Define what is train and what is test.
sales_train_validation['part'] = 'train'
test1['part'] = 'test1'
test2['part'] = 'test2'

In [9]:
#Concat
data = pd.concat([sales_train_validation, test1, test2], axis = 0)
#del sales_train_validation, test1, test2

#Verify null
print(data.isnull().sum())
print()
print(data.isnull().sum() / len(data))
print()
data.info()

id          0
item_id     0
dept_id     0
cat_id      0
store_id    0
state_id    0
day         0
demand      0
part        0
dtype: int64

id          0.0
item_id     0.0
dept_id     0.0
cat_id      0.0
store_id    0.0
state_id    0.0
day         0.0
demand      0.0
part        0.0
dtype: float64

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61742250 entries, 0 to 1707439
Data columns (total 9 columns):
id          object
item_id     object
dept_id     object
cat_id      object
store_id    object
state_id    object
day         object
demand      int16
part        object
dtypes: int16(1), object(8)
memory usage: 4.3+ GB


About the products:
The M5 dataset, generously made available by Walmart, involves the unit sales of various products sold in the USA, organized in the form of grouped time series. More specifically, the dataset involves the unit sales of 3,075 products, classified in 3 product categories (Hobbies, Foods, and Household) and 7 product departments, in which the above-mentioned categories are disaggregated. The products are sold across 10 stores, located in 3 States (CA, TX, and WI). In this respect, the bottom-level of the hierarchy, i.e., product-store unit sales, can be mapped either across product categories or geographical regions. More information can be found in the [M5 Participants Guide](https://mofc.unic.ac.cy/m5-competition/).

In [10]:
#Verify the numbers described above.

products = product_id['item_id'].value_counts(dropna = False) #3075 There is only 3049. Howeveer it is possible that M5's test database has these missing products.
departments = product_id['dept_id'].value_counts(dropna = False) #7 OK!
categories = product_id['cat_id'].value_counts(dropna = False) #3 OK!
stores = product_id['store_id'].value_counts(dropna = False) #10 OK!
States = product_id['state_id'].value_counts(dropna = False) #3 OK!

print('States:')
print('')
display(States)
print('')
print('stores:')
print('')
display(stores)
print('')
print('categories:')
print('')
display(categories)
print('')
print('departments:')
print('')
display(departments)
print('')
print('products:')
print('')
display(products)

States:



CA    12196
WI     9147
TX     9147
Name: state_id, dtype: int64


stores:



WI_2    3049
CA_2    3049
TX_2    3049
CA_3    3049
WI_3    3049
CA_1    3049
TX_3    3049
TX_1    3049
WI_1    3049
CA_4    3049
Name: store_id, dtype: int64


categories:



FOODS        14370
HOUSEHOLD    10470
HOBBIES       5650
Name: cat_id, dtype: int64


departments:



FOODS_3        8230
HOUSEHOLD_1    5320
HOUSEHOLD_2    5150
HOBBIES_1      4160
FOODS_2        3980
FOODS_1        2160
HOBBIES_2      1490
Name: dept_id, dtype: int64


products:



FOODS_1_097        10
HOUSEHOLD_2_376    10
FOODS_1_045        10
HOUSEHOLD_2_210    10
HOUSEHOLD_1_452    10
                   ..
FOODS_2_298        10
FOODS_3_434        10
FOODS_2_244        10
HOUSEHOLD_1_097    10
FOODS_3_721        10
Name: item_id, Length: 3049, dtype: int64

In [11]:
#Verify holidays
event_name_1 = calendar['event_name_1'].value_counts(dropna = False) #30
event_type_1 = calendar['event_type_1'].value_counts(dropna = False) #4
event_name_2 = calendar['event_name_2'].value_counts(dropna = False) #4
event_type_2 = calendar['event_type_2'].value_counts(dropna = False) #2


#It shoud be binary variables:
snap_CA = calendar['snap_CA'].value_counts(dropna = False) #49.28% are snap days
snap_TX = calendar['snap_TX'].value_counts(dropna = False)
snap_WI = calendar['snap_WI'].value_counts(dropna = False)

print('event_name_1:')
print('')
display(event_name_1)
print('')
print('event_type_1:')
print('')
display(event_type_1)
print('')
print('event_name_2:')
print('')
display(event_name_2)
print('')
print('event_type_2:')
print('')
display(event_type_2)
print('')
print('snap_CA:')
print('')
display(snap_CA)
print('')
print('snap_TX:')
print('')
display(snap_TX)
print('')
print('snap_WI:')
print('')
display(snap_WI)

event_name_1:



NaN                    1807
ValentinesDay             6
StPatricksDay             6
LentStart                 6
NBAFinalsEnd              6
Ramadan starts            6
PresidentsDay             6
Mother's day              6
LentWeek2                 6
Purim End                 6
MemorialDay               6
SuperBowl                 6
Pesach End                6
NBAFinalsStart            6
Cinco De Mayo             5
LaborDay                  5
VeteransDay               5
Thanksgiving              5
Chanukah End              5
Easter                    5
OrthodoxEaster            5
EidAlAdha                 5
Christmas                 5
OrthodoxChristmas         5
Halloween                 5
MartinLutherKingDay       5
NewYear                   5
IndependenceDay           5
ColumbusDay               5
Eid al-Fitr               5
Father's day              4
Name: event_name_1, dtype: int64


event_type_1:



NaN          1807
Religious      55
National       52
Cultural       37
Sporting       18
Name: event_type_1, dtype: int64


event_name_2:



NaN               1964
Father's day         2
OrthodoxEaster       1
Cinco De Mayo        1
Easter               1
Name: event_name_2, dtype: int64


event_type_2:



NaN          1964
Cultural        4
Religious       1
Name: event_type_2, dtype: int64


snap_CA:



0    1319
1     650
Name: snap_CA, dtype: int64


snap_TX:



0    1319
1     650
Name: snap_TX, dtype: int64


snap_WI:



0    1319
1     650
Name: snap_WI, dtype: int64

In [12]:
#Add information in our dataset.
#Merge with calendar.
data = pd.merge(data, calendar, how = 'left', left_on = ['day'], right_on = ['d'])
data.drop(['d'], inplace = True, axis = 1)

#Verify missing.
print(data.isnull().sum())
print()
print(data.isnull().sum() / len(data))

id                     0
item_id                0
dept_id                0
cat_id                 0
store_id               0
state_id               0
day                    0
demand                 0
part                   0
date                   0
wm_yr_wk               0
weekday                0
wday                   0
month                  0
year                   0
event_name_1    56558950
event_type_1    56558950
event_name_2    61559310
event_type_2    61559310
snap_CA                0
snap_TX                0
snap_WI                0
dtype: int64

id              0.000000
item_id         0.000000
dept_id         0.000000
cat_id          0.000000
store_id        0.000000
state_id        0.000000
day             0.000000
demand          0.000000
part            0.000000
date            0.000000
wm_yr_wk        0.000000
weekday         0.000000
wday            0.000000
month           0.000000
year            0.000000
event_name_1    0.916049
event_type_1    0.916049
event_name_

In [13]:
#Merge with sell_prices.
data = pd.merge(data, sell_prices, how = 'left', on = ['store_id', 'item_id', 'wm_yr_wk'])
#Verify missing.
print(data.isnull().sum())
print()
print(data.isnull().sum() / len(data))

print()
print(data.info())
display(data.head())

id                     0
item_id                0
dept_id                0
cat_id                 0
store_id               0
state_id               0
day                    0
demand                 0
part                   0
date                   0
wm_yr_wk               0
weekday                0
wday                   0
month                  0
year                   0
event_name_1    56558950
event_type_1    56558950
event_name_2    61559310
event_type_2    61559310
snap_CA                0
snap_TX                0
snap_WI                0
sell_price      12299413
dtype: int64

id              0.000000
item_id         0.000000
dept_id         0.000000
cat_id          0.000000
store_id        0.000000
state_id        0.000000
day             0.000000
demand          0.000000
part            0.000000
date            0.000000
wm_yr_wk        0.000000
weekday         0.000000
wday            0.000000
month           0.000000
year            0.000000
event_name_1    0.916049
event_type_

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  day  demand   part        date  ...  month  year  event_name_1  \
0       CA  d_1       0  train  2011-01-29  ...      1  2011           NaN   
1       CA  d_1       0  train  2011-01-29  ...      1  2011           NaN   
2       CA  d_1       0  train  2011-01-29  ...      1  2011           NaN   
3       CA  d_1       0  train  2011-01-29  ...      1  2011           NaN   
4       CA  d_1       0  train  2011-01-29  ...      1  2011           NaN   

   event_type_1  event_name_2 event_type_2 snap_CA snap_TX snap_WI  sell_price  
0           NaN           NaN          NaN       0       0       0         NaN  
1           NaN           NaN          NaN       0       0       0         NaN  
2           NaN           NaN          NaN       0       0       0         NaN  
3           NaN           NaN          NaN       0       0       0         NaN  
4           NaN           NaN          NaN       0       0       0         NaN  

[5 rows x 23 columns]

As we stated above, it is expected to have this many null values in sell_price.
We were expecting 1,757,059 weeks without price informantion, so 1,757,059 x 7 = 12,299,413

In [14]:
#make new variables
# time related variables
data['date'] = pd.to_datetime(data['date'])
data['year'] = data['date'].dt.year
data['month'] = data['date'].dt.month
data['week'] = data['date'].dt.week
data['day'] = data['date'].dt.day
data['dayofweek'] = data['date'].dt.dayofweek

gc.collect()
data['season'] = data.apply(lambda row: 'spring' if (row['month'] >= 3 and row['month'] <= 5)
                            else('summer' if (row['month'] >= 6 and row['month'] <= 8)
                            else('fall' if (row['month'] >= 9 and row['month'] <= 11)
                            else('winter'))), axis = 1)

data.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  day  demand   part       date  ...  event_type_1 event_name_2  \
0       CA   29       0  train 2011-01-29  ...           NaN          NaN   
1       CA   29       0  train 2011-01-29  ...           NaN          NaN   
2       CA   29       0  train 2011-01-29  ...           NaN          NaN   
3       CA   29       0  train 2011-01-29  ...           NaN          NaN   
4       CA   29       0  train 2011-01-29  ...           NaN          NaN   

   event_type_2  snap_CA  snap_TX snap_WI sell_price week dayofweek  season  
0           NaN        0        0       0        NaN    4         5  winter  
1           NaN        0        0       0        NaN    4         5  winter  
2           NaN        0        0       0        NaN    4         5  winter  
3           NaN        0        0       0        NaN    4         5  winter  
4           NaN        0        0       0        NaN    4         5  winter  

[5 rows x 26 columns]